sgkit/tests/test_association.py

```diff
@@ -0,0 +1,169 @@
+    data_vars = {}
+    data_vars["dosage"] = (["variant", "sample"], g.T)
+    for i in range(x.shape[1]):
+        data_vars[f"covar_{i}"] = (["sample"], x[:, i])
+    for i in range(len(ys)):
+        data_vars[f"trait_{i}"] = (["sample"], ys[i])
+    attrs = dict(beta=bg)
+    return xr.Dataset(data_vars, attrs=attrs)  # type: ignore[arg-type]
+
+
+@pytest.fixture  # type: ignore[misc]
+def ds() -> Dataset:
+    return _generate_test_dataset()
+
+
+def _sm_statistics(
+    ds: Dataset, i: int, add_intercept: bool
+) -> RegressionResultsWrapper:
+    X = []
+    # Make sure first independent variable is variant
+    X.append(ds["dosage"].values[i])
+    for v in [c for c in list(ds.keys()) if c.startswith("covar_")]:
+        X.append(ds[v].values)
+    if add_intercept:
+        X.append(np.ones(ds.dims["sample"]))
+    X = np.stack(X).T
+    y = ds[f"trait_{i}"].values
+
+    return sm.OLS(y, X, hasconst=True).fit()
+
+
+def _get_statistics(
+    ds: Dataset, add_intercept: bool, **kwargs: Any
+) -> Tuple[DataFrame, DataFrame]:
+    df_pred: List[Dict[str, Any]] = []
+    df_true: List[Dict[str, Any]] = []
+    for i in range(ds.dims["variant"]):
+        dsr = gwas_linear_regression(
+            ds,
+            dosage="dosage",
+            trait=f"trait_{i}",
+            add_intercept=add_intercept,
+            **kwargs,
+        )
+        res = _sm_statistics(ds, i, add_intercept)
+        df_pred.append(
+            dsr.to_dataframe()  # type: ignore[no-untyped-call]
+            .rename(columns=lambda c: c.replace("variant/", ""))
+            .iloc[i]
+            .to_dict()
+        )
+        df_true.append(dict(t_value=res.tvalues[0], p_value=res.pvalues[0]))
+    return pd.DataFrame(df_pred), pd.DataFrame(df_true)
+
+
+def test_linear_regression_statistics(ds):
+    def validate(dfp: DataFrame, dft: DataFrame) -> None:
+        print(dfp)
+        print(dft)
+
+        # Validate results at a higher level, looking only for recapitulation
+        # of more obvious inferences based on how the data was simulated
+        np.testing.assert_allclose(dfp["beta"], ds.attrs["beta"], atol=1e-3)
+        mid_idx = ds.dims["variant"] // 2
+        assert np.all(dfp["p_value"].iloc[:mid_idx] > 0.05)
+        assert np.all(dfp["p_value"].iloc[mid_idx:] < 0.05)
+
+        # Validate more precisely against statsmodels results
+        np.testing.assert_allclose(dfp["t_value"], dft["t_value"])
+        np.testing.assert_allclose(dfp["p_value"], dft["p_value"])
+
+    dfp, dft = _get_statistics(
+        ds, covariates=["covar_0", "covar_1", "covar_2"], add_intercept=True
+    )
+    validate(dfp, dft)
+
+    dfp, dft = _get_statistics(
+        ds.assign(covar_3=("sample", np.ones(ds.dims["sample"]))),
+        covariates=["covar_0", "covar_1", "covar_2", "covar_3"],
+        add_intercept=False,
+    )
+    validate(dfp, dft)
+
+
+def test_linear_regression_raise_on_no_covars(ds):
+    with pytest.raises(ValueError):
```

Comment:
afair you can add `matches="At least one covariate must be provided")` to makes sure you are asserting the right `ValueError`

Reply:
https://github.com/pystatgen/sgkit/pull/16/commits/273375e5d949ae3cd6b22572a6fccb2bd3064f0f